Add external temperature sensor support for Aqara E1 TRV

Elyasin opened the pull request:
## Proposed change
Adds support for external temperature sensor.

## Additional information
This is a continuation of https://github.com/zigpy/zha-device-handlers/pull/3974 (which itself is a continuation of https://github.com/zigpy/zha-device-handlers/pull/2802). The review comments have been considered in this PR. Additionally added two tests and did some refactoring.

## Checklist
- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works

ElyasinNtt committed:
Adds support for Aqara E1 TRV external temp sensor

ElyasinNtt committed:
Refactors for maintainability

ElyasinNtt committed:
Tests read forwarding and mapping for system mode

Elyasin committed:
Merge branch 'zigpy:dev' into dev

Elyasin committed:
Merge branch 'zigpy:dev' into dev

m00nyONE commented:
It would be very
Awesome if someone finally will merge this PR after almost 2 years 😂 it's the only thing that is holding me from switching from z2m to zha

OskarKinomoto commented:
Is there a way to help you so it is possible to merge it?

Elyasin committed:
Merge branch 'zigpy:dev' into dev

Elyasin committed:
Merge branch 'zigpy:dev' into dev

cl445 commented:
Hi @Elyasin,

I've opened a complementary PR #4604 that focuses on different aspects of the AGL001 quirk:

**My PR (#4604) adds:**
- Heartbeat parsing (attribute 0x00F7) for device/local temperature, battery, preset, valve alarm
- `LocalDeviceTemperatureCluster` for internal TRV temperature
- Type-safe enums (`SystemMode`, `Preset`, `SensorMode`)
- Code quality improvements (`t.Bool`, manufacturer code constant)

**Your PR (#4463) adds:**
- External sensor mode switching
- External temperature input

These PRs are **complementary and non-conflicting**. They could be merged independently or combined. My PR intentionally leaves the external sensor functionality as TODO, referencing your work.

Would be great to see both merged - together they would bring ZHA's AGL001 support close to feature parity with zigbee2mqtt! 🎉

fridgecow commented:
> These PRs are **complementary and non-conflicting**. They could be merged independently or combined. My PR intentionally leaves the external sensor functionality as TODO, referencing your work.

They may be complementary, but they are also definitely conflicting. Whoever merges second is going to have a few merge conflicts to deal with.

GamerClassN7 commented:
Any news on this MR ?

Elyasin committed:
Merge branch 'zigpy:dev' into dev

Elyasin committed:
Address review comments: improve code quality and documentation
- Add modulo arithmetic to integrity calculation for overflow prevention
- Rename temperatureBuf to temperature_buf (PEP 8 compliance)
- Replace magic numbers with named constants in Constants class
- Fix return type annotations (aqara_header, _float_to_hex, _build_sensor_mode_params)
- Fix parameter type annotation (params: bytes)
- Change SENSOR_ID from mutable bytearray to immutable bytes
- Add comprehensive docstrings to Constants class and methods
- Remove redundant @pytest.mark.asyncio decorator
- Clarify temperature value comment
All tests pass successfully.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+    def aqara_header(self, counter: int, params: bytearray, action: int) -> bytearray:
+        """Create Aqara header for setting external sensor."""
+        header = bytes([0xAA, 0x71, len(params) + 3, 0x44, counter])
+        integrity = 512 - sum(header)
```
✅ Fixed. Applied modulo arithmetic: `integrity = (512 - sum(header)) % 256`

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+
+                params = bytearray(Constants.SENSOR_ID)
+                params += bytes([0x00, 0x01, 0x00, 0x55])
+                params += temperatureBuf
```
✅ Fixed. Renamed `temperatureBuf` to `temperature_buf` to follow PEP 8 snake_case convention.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+                attrs[self.AttributeDefs.sensor_attr.name] = (
+                    self.aqara_header(0x13, params2, 0x02 if is_external else 0x04)
+                    + params2
+                )
```
✅ Fixed. Added named constants to `Constants` class:
- `ACTION_SET_EXTERNAL_TEMP = 0x05`
- `ACTION_SET_INTERNAL_SENSOR = 0x04`
- `ACTION_SET_EXTERNAL_SENSOR = 0x02`
- `HEADER_ACTION_SENSOR_TEMP = 0x12`
- `HEADER_ACTION_SENSOR_MODE = 0x13`

All usages updated to reference these constants.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+                )
+
+                params = bytearray(Constants.SENSOR_ID)
+                params += bytes([0x00, 0x01, 0x00, 0x55])
```
✅ Fixed. Replaced `bytes([0x00, 0x01, 0x00, 0x55])` with `Constants.SENSOR_ID_SUFFIX`.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+        header = bytes([0xAA, 0x71, len(params) + 3, 0x44, counter])
+        integrity = 512 - sum(header)
+
+        return header + bytes([integrity, action, 0x41, len(params)])
```
✅ Fixed. Changed return type from `bytearray` to `bytes` to match actual return value.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+XIAOMI_SYSTEM_MODE_MAP = {
+    0: Thermostat.SystemMode.Off,
+    1: Thermostat.SystemMode.Heat,
+}
+
+
+class Constants:
+    """Constants specific for Aqara E1 TRV."""
+
+    SENSOR_ID = bytearray.fromhex("00158d00019d1b98")
+    SENSOR_ID_SUFFIX = bytes([0x00, 0x01, 0x00, 0x55])
+    INTERNAL_SENSOR_ACTION_CODES = [bytes([0x3D, 0x05]), bytes([0x3D, 0x04])]
+    EXTERNAL_SENSOR_ACTION_CODES = [bytes([0x3D, 0x04]), bytes([0x3D, 0x05])]
+    INTERNAL_SENSOR_PADDING = bytes(12)
+    EXTERNAL_SENSOR_DATA_BLOCK_1 = bytes(
+        [0x13, 0x0A, 0x02, 0x00, 0x00, 0x64, 0x04, 0xCE, 0xC2, 0xB6, 0xC8]
+    )
+    EXTERNAL_SENSOR_DATA_BLOCK_2 = bytes(
+        [
+            0x16,
+            0x0A,
+            0x02,
+            0x0A,
+            0xC9,
+            0xE8,
+            0xB1,
+            0xB8,
+            0xD4,
+            0xDA,
+            0xCF,
+            0xDF,
+            0xC0,
+            0xEB,
+        ]
+    )
+    EXTERNAL_SENSOR_TRAILING_1 = bytes([0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x3D])
+    EXTERNAL_SENSOR_TRAILING_2 = bytes([0x64])
+    EXTERNAL_SENSOR_TRAILING_3 = bytes([0x65])
+    EXTERNAL_SENSOR_PARAMS2_SUFFIX = bytes([0x08, 0x00, 0x07, 0xFD])
+    EXTERNAL_SENSOR_TRAILING_4 = bytes([0x04])
```
✅ Fixed. Added comprehensive docstring explaining protocol-specific byte sequences for Aqara E1 TRV proprietary commands.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+
+        return header + bytes([integrity, action, 0x41, len(params)])
+
+    def _float_to_hex(self, f):
```
✅ Fixed. Added return type annotation: `-> str`

Elyasin commented on tests/test_aqara_trv.py:
```diff
@@ -0,0 +1,232 @@
+
+
+@pytest.mark.parametrize("quirk", (AGL001,))
+@pytest.mark.asyncio
```
✅ Fixed. Removed redundant `@pytest.mark.asyncio` decorator.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+class Constants:
+    """Constants specific for Aqara E1 TRV."""
+
+    SENSOR_ID = bytearray.fromhex("00158d00019d1b98")
```
⏸️ Not addressed yet. In progress.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+            )
+        super()._update_attribute(attrid, value)
+
+    def aqara_header(self, counter: int, params: bytearray, action: int) -> bytearray:
```
✅ Fixed. Changed parameter type from `bytearray` to `bytes` for consistency.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+class Constants:
+    """Constants specific for Aqara E1 TRV."""
+
+    SENSOR_ID = bytearray.fromhex("00158d00019d1b98")
```
✅ Fixed. Changed from `bytearray.fromhex()` to `bytes.fromhex()` to prevent accidental mutations.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+            attr_def = self.find_attribute(attr)
+
+            if attr_def and attr_def.id == self.AttributeDefs.sensor_temp.id:
+                # set external sensor temp. this function expect value to be passed multiplied by 100
```
✅ Fixed. Updated comment to: "value is expected in degrees Celsius (e.g. 25.0 for 25°C)"

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+
+    def _build_sensor_mode_params(
+        self, is_external: bool, device: bytes, timestamp: bytes
+    ):
```
✅ Fixed. Added return type annotation: `-> tuple[bytes, bytes]`

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+        super()._update_attribute(attrid, value)
+
+    def aqara_header(self, counter: int, params: bytearray, action: int) -> bytearray:
+        """Create Aqara header for setting external sensor."""
```
✅ Fixed. Added comprehensive docstring with detailed parameter descriptions and examples of action codes used.

Elyasin commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,480 +1,653 @@
+                    f"{self.endpoint.device.ieee}".replace(":", "")
+                )
+
+                timestamp = bytes(reversed(t.uint32_t(int(time.time())).serialize()))
```
The review comment is incorrect. `time.time()` returns the time since epoch and not in system's local timezone. Source: Python 3 documentation.

cl445 commented:
Hi @Elyasin,

I've been working on a separate PR (#4604) that migrates the Aqara E1 TRV quirk to the **QuirkBuilder v2 API**, which is now the recommended approach. My PR includes:

- QuirkBuilder v2 migration with complete entity definitions
- Heartbeat parsing for temperature, battery, preset, valve alarm
- `running_state` simulation (fixes "unknown" HVAC action)
- Firmware version extraction
- Calibrate button

However, I intentionally did **not** include external sensor support, as I didn't want to duplicate your work here.

**Proposal:** Would you be interested in rebasing your external sensor implementation on top of my v2 quirk once #4604 is merged? This would:

1. Give both PRs a better chance of being accepted (v2 is the new standard)
2. Keep your work and proper attribution intact
3. Result in a complete, modern quirk for the Aqara E1 TRV

The v2 API is quite different from v1, but the core logic of your `write_attributes` override and the protocol constants should transfer well. I'd be happy to help with the migration if needed.

What do you think?

Esp32-zapper commented:
My bedroom temp sensor reads 13⁰C even if the Thermostat sensor reads 22⁰C, all because of a curtain, so please hurry up.
Tell wife we should have other curtains?
Well, I'm not suicidal, just really, really cold?? 😂🤣

Antoninj commented:
At this point it feels like the universe is against those changes.

Esp32-zapper commented:
> At this point it feels like the universe is against those changes.

What a pitty!

Elyasin committed:
Merge remote-tracking branch 'upstream' into dev

Esp32-zapper commented:
Didn't understand nada of that.
Does it mean we dare to hope?

Elyasin committed:
fix: map zcl / xiaomi values correctly

Elyasin commented:
> Didn't understand nada of that. Does it mean we dare to hope?

There was a merge conflict, which I tried to resolve. I need to keep the PR up to date with the destination branch, so that it can be reviewed properly. I hope that's ok now.

Elyasin committed:
Merge branch 'zigpy:dev' into dev

Elyasin committed:
Merge remote-tracking branch 'upstream' into dev

ovcharove commented:
Hi. Everything worked on the first test connection. After reinstalling Haos and reconnecting the thermostat, an error appears when trying to switch to an external sensor: "ValueError: Record Attribute(attrid=0xFFF2, value=TypeValue(type=LVBytes, value=b'\xaaq1D\x12^\x02A.j2\xc7\x94=\x04T\xefD\x10\x00\x9ac\x7f\x00\x15\x8d\x00\x01\x9d\x1b\x98\x00\x01\x00U\x13\n\x02\x00\x00d\x04\xce\xc2\xb6\xc8\x00\x00\x00\x00\x00\x01=de')) is too large to fit in a single request: 59 > 50 bytes"

TheJulianJES commented:
@ovcharove Sorry about that! That was (unintentionally) introduced by https://github.com/zigpy/zigpy/pull/1823 for this device. https://github.com/zigpy/zigpy/issues/1862 is the issue for it and we'll have a proper fix in zigpy shortly.

---

Also, after going through some big ZHA refactors, we're (finally) getting to processing the backlog of all quirks PRs next. I'll have the (internal) review posted here already but note that we may be able to address all of the (valid) issues ourselves, so don't stress on keeping this PR up-to-date all the time. It's not a blocker for it to be reviewed. I'm aware of all the Aqara E1 TRV PRs (e.g. also https://github.com/zigpy/zha-device-handlers/pull/4604 + https://github.com/zigpy/zha-device-handlers/pull/4657). We'll likely merge the quirks v2 refactors first.

TheJulianJES commented on zhaquirks/xiaomi/aqara/thermostat_agl001.py:
```diff
@@ -1,484 +1,717 @@
+                timestamp = bytes(reversed(t.uint32_t(int(time.time())).serialize()))
+
+                is_external = bool(value)
+                params1, params2 = self._build_sensor_mode_params(
```
Reported in https://github.com/zigpy/zigpy/issues/1862 and we'll have a fix for this shortly. This seems to be the only device needing more than 50 bytes when writing, so we missed this.
